Polyhedron/demo/Polyhedron/Plugins/PMP/Smoothing_plugin.cpp

```diff
@@ -0,0 +1,360 @@
+
+      poly_item->invalidateOpenGLBuffers();
+      poly_item->itemChanged();
+    }
+
+    else if(selection_item)
+    {
+      smooth_angles(selection_item->selected_facets, pmesh,
+                    parameters::number_of_iterations(nb_iter)
+                    .do_project(projection));
+
+      selection_item->poly_item_changed();
+      selection_item->changed_with_poly_item();
+    }
+
+    QApplication::restoreOverrideCursor();
+   }
+
+  void on_area_smoothing_clicked()
+  {
+    const Scene_interface::Item_id index = scene->mainSelectionIndex();
+    Scene_polyhedron_item* poly_item = qobject_cast<Scene_polyhedron_item*>(scene->item(index));
+    Scene_polyhedron_selection_item* selection_item = qobject_cast<Scene_polyhedron_selection_item*>(scene->item(index));
+    Polyhedron& pmesh = (poly_item != NULL) ? * poly_item->polyhedron() : * selection_item->polyhedron();
+
+    unsigned int nb_iter = ui_widget.areas_iter_spinBox->value();
+    bool projection = ui_widget.projection_checkBox->isChecked();
+    const double precision = ui_widget.precision_spinBox->value();
+    QApplication::setOverrideCursor(Qt::WaitCursor);
+
+    if(poly_item)
+    {
+      smooth_areas(pmesh, parameters::number_of_iterations(nb_iter)
+                                      .do_project(projection)
+                                      .gradient_descent_precision(precision));
+      poly_item->invalidateOpenGLBuffers();
+      poly_item->itemChanged();
+    }
+
+    else if(selection_item)
+    {
+      smooth_areas(selection_item->selected_facets, pmesh, parameters::number_of_iterations(nb_iter)
+                                                          .do_project(projection)
+                                                          .gradient_descent_precision(precision));
+      selection_item->poly_item_changed();
+      selection_item->changed_with_poly_item();
+    }
+    else
+    {
+      std::cerr<< "Something's gone wrong.\n";
+    }
+
+    QApplication::restoreOverrideCursor();
+   }
+
+  void on_curvature_flow_clicked()
+  {
+    const Scene_interface::Item_id index = scene->mainSelectionIndex();
+    Scene_polyhedron_item* poly_item = qobject_cast<Scene_polyhedron_item*>(scene->item(index));
+    Scene_polyhedron_selection_item* selection_item = qobject_cast<Scene_polyhedron_selection_item*>(scene->item(index));
+    Polyhedron& pmesh = (poly_item != NULL) ? * poly_item->polyhedron() : * selection_item->polyhedron();
+
+    int index_id = scene->item_id(poly_item);
+    const double time_step = ui_widget.time_step_spinBox->value();
+    const unsigned int nb_iter = ui_widget.curvature_iter_spinBox->value();
+
+    const bool use_constrained_vertex_map = ui_widget.border_button->isChecked()
+                                            && !CGAL::is_closed(pmesh);
+
+    QApplication::setOverrideCursor(Qt::WaitCursor);
+
+    // explicit scheme
+    if(ui_widget.explicit_checkBox->isChecked())
```

Comment:
I wonder why in case we do not use the explicit scheme we are using some internal functions with some caching and not simply pass false to `use_explicit_scheme()`.
@janetournois any idea?

Reply:
I'm trying to figure this out

Reply:
in the demo, this makes possible to compute the stiffness matrix, and then use it without recomputing it, on the cached item you mention below.
I think we can remove that feature. It may be faster in some cases for demo purposes, but it introduces confusion in the plugin.
I will do it